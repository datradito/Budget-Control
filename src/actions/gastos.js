import uuid from 'uuid';
import database from '../firebase/firebase';


//EXPORTO POR NOMBRE, LUEGO ELIJO AL IMPORTAR
//ADD_GASTO - action generator
/* const funcAction =  ({} = {}) => ( {} ) */

export const addGasto =  (gasto) => ({
    type: 'ADD_GASTO',
    gasto
});

//ACCION COMPATIBLE PARA ENVIAR CON THUNK
export const startAddGasto = (( gastoDato = {} ) => {
    return (dispatch) => {
        const {
            descrip= '',
            note= '',
            importe= 0, 
            creadoAt= 0,
        } = gastoDato;

        const gasto = { descrip, note, importe, creadoAt };

        return database.ref('gastos').push(gasto)
        //pasando ref tengo acceso al .key desde firebase
        .then((ref) => {
            dispatch(addGasto({
                id: ref.key,
                ...gasto
            }));
        });
    };
});

//REMOVE_GASTO
export const removeGasto = ( { id } ) => ({
        type: 'REMOVE_GASTO',
        id
    
    });
//EDIT_GASTO
//no asigno default xq sino no tengo id/updates no hago nada en el reducer
export const editGasto = ( id, updates ) => ({
        type: 'EDIT_GASTO',
        id,
        updates
    }
);
    


//CODIGO PARA thunk
// ( gasto ) => ({
//         type: 'ADD_GASTO',
//         //gasto
//         //QUITO OBJETO XQ PASO TODO A FIREBASE
//         gasto: {
//             //id generado auto by uuid@3.1.0
//             id: uuid(),
//             //key:value iguales entonces simplifico
//             descrip,
//             note,
//             importe,
//             creadoAt
//         }
//     });



