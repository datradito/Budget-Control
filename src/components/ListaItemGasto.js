import React from 'react';
import { Link } from 'react-router-dom';

//STATELESS FUNCTIONAL COMPONENT:
const ListaItemGasto = ({ id, descrip, importe, creadoAt}) => (
    <div>
        <Link to= {`/edit/${ id }`}>
            <h3>Descripcion: {descrip} </h3> 
        </Link>

        {/*como las props me vienen destructuradas no utilizo prop.importe x ej*/}
        <p> Importe: ${importe} - Fecha: {creadoAt} milisegundos</p>
        <p>ID: {id} </p>



    </div>
);

export default ListaItemGasto;

