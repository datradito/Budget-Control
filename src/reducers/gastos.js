//Reducer Gastos

//Variables DEFAULT
//el default es un array, no un obj individual.
const gastosReducerDefault = [];

export default (state = gastosReducerDefault, action) => {
    switch (action.type) {
        case 'ADD_GASTO':
        //ES6 spread
        //[...state, {}] --->> el spread crea un nuevo array en memoria con el AGREGADO del objeto q envia dispatch(action), NO SE MODIFICA EL STATE!
            return [
                ...state,
                action.gasto
            ]
 
        case 'REMOVE_GASTO':
            //DEVUELVO SOLO LOS ELEMENTOS DISTINTOS AL id pasado a removeGasto 
            /* 
                id!==id -->> son distintos los idÂ¿?
                si true (son distintos) -->> sigo con el sgte item del array, no modifico nada
                si false (son iguales) -->> filtro el item, no lo incorporo al nuevo array.
            */
            return state.filter(({ id }) => id!== action.id);
            //IDEM
            /*return state.filter(({ id }) => {
                return id !== action.id
            }); */
        
        case 'EDIT_GASTO':
            //mapeo todas las props del obj gastos
            return state.map((gasto) => {
                //si encuentro el id a modificar
                if (gasto.id === action.id) {
                    //Entonces hago OBJ SPREAD OPERATOR para devolver el obj modificado
                    return {
                        ...gasto,
                        ...action.updates
                    };
                } else {
                    return gasto;
                };
            });

        default:
            return state;
    }
};