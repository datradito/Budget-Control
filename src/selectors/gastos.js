import moment from 'moment';

//Envio Gastos filtrados a la consola
//gastos(obj gastos, obj filters)
//obj filters : {text, sortBy, fechaInicio, fechaFin} 

export default (gastos, { text, sortBy, fechaInicio, fechaFin}) => {
    //.filter() devuelve un array con los match q dieron true para cada filtro:fechaInicioMatch, fechaFinMatch, textMatch.
    return gastos.filter((gasto) => {
        //Logica para filtrar dentro de fechaInicio-fechaFin
        const creadoAtMoment = moment(gasto.creadoAt);
        const fechaInicioMatch = fechaInicio ? fechaInicio.isSameOrBefore(creadoAtMoment, 'day') : true;            
        const fechaFinMatch = fechaFin ? fechaFin.isSameOrAfter(creadoAtMoment, 'day') : true;
        //filtro la descripcion si text esta incluida en la cadena , ver .includes()
        const textMatch = gasto.descrip.toLowerCase().includes(text.toLowerCase());

        return fechaInicioMatch && fechaFinMatch && textMatch;
        //ordeno por fecha de menor a mayor, e importe de mayor a menor. Ver  array.sort()
    }).sort((a, b ) => {
        //ordeno fecha
        if (sortBy === 'fecha') {
            return a.creadoAt < b.creadoAt ? 1 : -1; //ver compare en .sort()
        }
        //ordeno importe
        else if(sortBy === 'importe'){
             return a.importe < b.importe ? 1 : -1;
        }
    });
};

