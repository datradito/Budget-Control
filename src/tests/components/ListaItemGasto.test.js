import React from 'react';
import { shallow } from 'enzyme';
import gastos from '../fixtures/gastos';
import ListaItemGasto from '../../components/ListaItemGasto';


test('Deberia render ListaItemGasto con gastos', () => {
    const wrapper = shallow(<ListaItemGasto {...gastos[0]}/>);
    expect(wrapper).toMatchSnapshot();
});