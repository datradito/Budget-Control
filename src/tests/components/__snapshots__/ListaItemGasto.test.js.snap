// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Deberia render ListaItemGasto con gastos 1`] = `
<div>
  <Link
    to="/edit/1"
  >
    <h3>
      Descripcion: 
      Gas
       
    </h3>
  </Link>
  <p>
     Importe: $
    195
     - Fecha: 
    0
     milisegundos
  </p>
  <p>
    ID: 
    1
     
  </p>
</div>
`;
